docs/architecture.md

```diff
@@ -8,16 +8,109 @@ This whole application was built off of [Node.js](https://nodejs.org/en/)!
+#### `Route`
+
+A `Route` is defined by its uri path, its query parameters in the uri path, and the HTTP verb used to access it:
+
+```json
+{
+    "requestType": "POST",
+    "uri": "/api/sponsor/",
+}
+```
+
+#### `Role`
+
+A `Role` is a collection of `Routes`, and a unique name (such as `hacker`, or `sponsor`, or `my-custom-role`). This collection of `Routes` should approximately encapsulate the set of actions that a user of the api should be able to do. For example, a `Hacker` might want to have access to `GET /api/hacker/507f1f77bcf86cd799439011` (get their own information), or `POST /api/hacker/` (create their hacker document):
+
+```json
+{
+    "accountId": "507f191e810c19729de860ea",
+    "roles": ["54759eb3c090d83494e2d804", "51325eb3c090d83494e2d702"],
+}
+```
+
+_note here that `roles` is a list of `ObjectIds`, since the role is also a document in the database._
+
+#### `RoleBinding`
+
+A `RoleBinding` is a mapping between an `account` and a `Role`. In this way, we can say that an account has a certain set of allowed actions, defined by the `Role`. An `account` can have a `RoleBinding` to multiple `Roles`.
+
+### How to write a `Route` document
+
+In order to encapsulate permissions based on the actual `params` of a given route (such as the resource id in `/api/hacker/507f1f77bcf86cd799439011`), we use two placeholders: `:self`, and `:all`. These two placeholders replace the location of the given parameter in the route; in the example above, we would translate the uri to either `/api/hacker/:self`, or `/api/hacker/:all`.
+
+#### `:self`
+
+The `self` placeholder provides the account that hits this `Route` access to only the resource that is related to this particular accountId. For example, if I wanted to hit `GET /api/hacker/507f1f77bcf86cd799439011`, then the hacker whose `id` was `507f1f77bcf86cd799439011` would have to have an `accountId` that is equal to the `id` of the account accessing this resource.
+
+#### `:all`
+
+The `all` placeholder provides the account that hits this `Route` access to any resource for a given id.
+
+#### Chaining :self and :all
+
+It is possible to chain `:self` and `:all` together if need-be.
+
+### Adding Authentication and Authorization to your middleware
+
+We've tried to make this as simple and abstract as possible. To add `Authentication`, you just need to import the `Auth` module file, and then insert into the route definition:
+
+```javascript
+const Middleware = {
+    Auth: require("../../middlewares/auth.middleware")
+};
+...
+hackerRouter.route("/").post(
+    Middleware.Auth.ensureAuthenticated(),
+    // Whatever middleware is after this will run only if the user is authenticated.
+);
+```
+
+To add `Authorization`, you will need to import the `Auth` module file, and then insert into the route definition, just like `Authentication`. However, if there are route parameters, you will need to also provide as arguments the functions required to access the given parameters. An example is below:
```

Comment:
In particular, the functions needs to take in the parameter, and return an object. For authorization to succeed, the object's _id needs to match the user id, or the object's accountId needs to match the user's id. For practical purposes, the function just needs take the route parameter and return the object that the route is trying to access.

Also, if there are multiple route parameters, the functions need to be in the same order as the route parameters. For example, if fn1 corresponds to :id1, and fn2 corresponds to :id2, and the route is /api/abc/:id1/:id2, the functions need to be [fn1, fn2]. 

It's also important to note that even if there is only one route parameter, the argument for ensureAuthenticated needs to be in the form on an array of that one function. Ex: [fn1].